File: paddler_client/src/client_management.rs
use futures_util::StreamExt;
use paddler_messaging::agent_controller_pool_snapshot::AgentControllerPoolSnapshot;
use paddler_messaging::agent_desired_state::AgentDesiredState;
use paddler_messaging::balancer_desired_state::BalancerDesiredState;
use paddler_messaging::buffered_request_manager_snapshot::BufferedRequestManagerSnapshot;
use paddler_messaging::chat_template::ChatTemplate;
use paddler_messaging::model_metadata::ModelMetadata;
use serde_json::from_str;
use tokio_util::sync::CancellationToken;
use url::Url;

use crate::agents_stream::AgentsStream;
use crate::buffered_requests_stream::BufferedRequestsStream;
use crate::error::Result;
use crate::http_client::HttpClient;
use crate::reports_health::ReportsHealth;
use crate::stream::sse::Sse;

#[derive(Clone)]
pub struct ClientManagement {
    http_client: HttpClient,
}

impl ClientManagement {
    #[must_use]
    pub fn new(url: Url) -> Self {
        Self {
            http_client: HttpClient::new(url),
        }
    }

    pub async fn get_agents(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<AgentControllerPoolSnapshot> {
        self.http_client
            .get_json(cancellation_token, "/api/v1/agents")
            .await
    }

    pub async fn get_balancer_desired_state(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<BalancerDesiredState> {
        self.http_client
            .get_json(cancellation_token, "/api/v1/balancer_desired_state")
            .await
    }

    pub async fn get_balancer_applicable_state(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<Option<AgentDesiredState>> {
        self.http_client
            .get_json(cancellation_token, "/api/v1/balancer_applicable_state")
            .await
    }

    pub async fn put_balancer_desired_state(
        &self,
        cancellation_token: CancellationToken,
        state: &BalancerDesiredState,
    ) -> Result<()> {
        self.http_client
            .put_json(cancellation_token, "/api/v1/balancer_desired_state", state)
            .await?;

        Ok(())
    }

    pub async fn get_buffered_requests(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<BufferedRequestManagerSnapshot> {
        self.http_client
            .get_json(cancellation_token, "/api/v1/buffered_requests")
            .await
    }

    pub async fn get_agents_stream(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<AgentsStream> {
        let response = self
            .http_client
            .get(cancellation_token.clone(), "/api/v1/agents/stream")
            .await?;

        let stream = Sse::from_response(cancellation_token, response)
            .map(|result| result.and_then(|data| from_str(&data).map_err(Into::into)));

        Ok(Box::pin(stream))
    }

    pub async fn get_buffered_requests_stream(
        &self,
        cancellation_token: CancellationToken,
    ) -> Result<BufferedRequestsStream> {
        let response = self
            .http_client
            .get(
                cancellation_token.clone(),
                "/api/v1/buffered_requests/stream",
            )
            .await?;

        let stream = Sse::from_response(cancellation_token, response)
            .map(|result| result.and_then(|data| from_str(&data).map_err(Into::into)));

        Ok(Box::pin(stream))
    }

    pub async fn get_chat_template_override(
        &self,
        cancellation_token: CancellationToken,
        agent_id: &str,
    ) -> Result<Option<ChatTemplate>> {
        self.http_client
            .get_json(
                cancellation_token,
                &format!("/api/v1/agent/{agent_id}/chat_template_override"),
            )
            .await
    }

    pub async fn get_model_metadata(
        &self,
        cancellation_token: CancellationToken,
        agent_id: &str,
    ) -> Result<Option<ModelMetadata>> {
        self.http_client
            .get_json(
                cancellation_token,
                &format!("/api/v1/agent/{agent_id}/model_metadata"),
            )
            .await
    }

    pub async fn get_metrics(&self, cancellation_token: CancellationToken) -> Result<String> {
        self.http_client
            .get_text(cancellation_token, "/metrics")
            .await
    }
}

impl ReportsHealth for ClientManagement {
    fn http_client(&self) -> &HttpClient {
        &self.http_client
    }
}

File: paddler_client_python/paddler_client/client_management.py
from __future__ import annotations

from typing import TYPE_CHECKING, Self

import httpx

from paddler_client.agent_controller_pool_snapshot import (
    AgentControllerPoolSnapshot,
)
from paddler_client.balancer_desired_state import BalancerDesiredState
from paddler_client.buffered_request_manager_snapshot import (
    BufferedRequestManagerSnapshot,
)
from paddler_client.chat_template import ChatTemplate
from paddler_client.error import HttpError
from paddler_client.model_metadata import ModelMetadata
from paddler_client.stream_sse import stream_sse

if TYPE_CHECKING:
    from collections.abc import AsyncIterator


class ClientManagement:
    def __init__(
        self,
        url: str,
        http_client: httpx.AsyncClient | None = None,
    ) -> None:
        self._url = url.rstrip("/")
        self._http_client = (
            http_client if http_client is not None else httpx.AsyncClient()
        )

    async def get_health(self) -> str:
        response = await self._http_client.get(f"{self._url}/health")

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        return response.text

    async def get_agents(self) -> AgentControllerPoolSnapshot:
        response = await self._http_client.get(
            f"{self._url}/api/v1/agents",
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        return AgentControllerPoolSnapshot.model_validate_json(
            response.content,
        )

    async def agents_stream(
        self,
    ) -> AsyncIterator[AgentControllerPoolSnapshot]:
        url = f"{self._url}/api/v1/agents/stream"

        async with self._http_client.stream("GET", url) as response:
            async for data in stream_sse(response):
                yield AgentControllerPoolSnapshot.model_validate_json(data)

    async def get_balancer_desired_state(self) -> BalancerDesiredState:
        response = await self._http_client.get(
            f"{self._url}/api/v1/balancer_desired_state",
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        return BalancerDesiredState.model_validate_json(response.content)

    async def put_balancer_desired_state(
        self,
        state: BalancerDesiredState,
    ) -> None:
        response = await self._http_client.put(
            f"{self._url}/api/v1/balancer_desired_state",
            content=state.model_dump_json(
                exclude_none=True,
                by_alias=True,
            ),
            headers={"Content-Type": "application/json"},
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

    async def get_buffered_requests(
        self,
    ) -> BufferedRequestManagerSnapshot:
        response = await self._http_client.get(
            f"{self._url}/api/v1/buffered_requests",
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        return BufferedRequestManagerSnapshot.model_validate_json(
            response.content,
        )

    async def buffered_requests_stream(
        self,
    ) -> AsyncIterator[BufferedRequestManagerSnapshot]:
        url = f"{self._url}/api/v1/buffered_requests/stream"

        async with self._http_client.stream("GET", url) as response:
            async for data in stream_sse(response):
                yield BufferedRequestManagerSnapshot.model_validate_json(
                    data,
                )

    async def get_chat_template_override(
        self,
        agent_id: str,
    ) -> ChatTemplate | None:
        response = await self._http_client.get(
            f"{self._url}/api/v1/agent/{agent_id}/chat_template_override",
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        if not response.content or response.text == "null":
            return None

        return ChatTemplate.model_validate_json(response.content)

    async def get_model_metadata(
        self,
        agent_id: str,
    ) -> ModelMetadata | None:
        response = await self._http_client.get(
            f"{self._url}/api/v1/agent/{agent_id}/model_metadata",
        )

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        if not response.content or response.text == "null":
            return None

        return ModelMetadata.model_validate_json(response.content)

    async def get_metrics(self) -> str:
        response = await self._http_client.get(f"{self._url}/metrics")

        if not response.is_success:
            raise HttpError(response.status_code, response.text)

        return response.text

    async def close(self) -> None:
        await self._http_client.aclose()

    async def __aenter__(self) -> Self:
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()
